airgun/widgets.py

```diff
@@ -30,22 +40,27 @@ def assign_resource(self, values):
         for value in values:
             self._filter_value(value)
             self.browser.click(
-                self.browser.element(self.LIST_FROM.locator % value))
+                self.browser.element(self.ITEM_FROM.locator % value))
 
     def unassign_resource(self, values):
         for value in values:
             self._filter_value(value)
             self.browser.click(
-                self.browser.element(self.LIST_TO.locator % value))
+                self.browser.element(self.ITEM_TO.locator % value))
 
-    def fill(self, dict_values):
-        if dict_values['operation'] == 'Add':
-            self.assign_resource(dict_values['values'])
-        if dict_values['operation'] == 'Remove':
-            self.unassign_resource(dict_values['values'])
+    def fill(self, values):
+        if values['operation'] == 'Add':
+            self.assign_resource(values['values'])
+        if values['operation'] == 'Remove':
+            self.unassign_resource(values['values'])
 
     def read(self):
-        do_not_read_this_widget()
+            return {
+                'free': [
+                    el.text for el in self.browser.elements(self.LIST_FROM)],
```

Comment:
Do not use **any** element attributes or methods directly unless they are missing in `browser`. use browser.text(el) instead, browser contains some hacks and workarounds for some known issues.

Reply:
I know about `browser.text`, but it is not applicable here. I did exactly what I needed, to return list of all found elements by pattern and then got text for each of them.

Reply:
WT is designed to have control over all possible things, so it is better to use the browser methods. If you do .text on the element, it does query selenium anyway :)